### PR TITLE
[XLA] vectorize row reduction for even row size (new version of PR #37260)

### PR DESCRIPTION
**Pull request by nouiz**
This is a new version of #37260 that is rabased and include the fix to the DCHECK and convert it to a CHECK:
https://github.com/tensorflow/tensorflow/pull/37260#pullrequestreview-385503035
It also include the other small comments.
@cheshire 
@akuegel 

**Original description:**

This depend on LLVM commit: https://reviews.llvm.org/D74444 that is recently merged in XLA.

This PR vectorize row reduction for even row size.
For odd row size, every other row is unaligned. So would need a more complicated fix.
On a reduction of size 32x131070, I have no speed up for a max reduction, but I have 1.21x speed on the sum+pw reduction from the softmax in float16.

This was benchmarked on the row major "Softmax" TF operations.
The optimized graph contain 3 main operations: reduce_8 (the max reduction), fusion_1 (the exponential and sum reduction), fusion (the post normalization pw operations, ignored).
This was benchmarked with a matrix of input shape (32x131070) with float16 dtype:

| Kernel     | Before PR | After PR | Speed up |
| ---------- | ---------- | ---------- | ---------- |
| fusion_1  | 39.827us  | 33.843us | 1.18x |
| reduce_8 | 22.112us  | 22.047us | 1x |

Here is the not optimized HLO:
```
HloModule cluster_0
%max_ {
  %x = f16[] parameter(0)
  %y = f16[] parameter(1)
  ROOT %maximum = f16[] maximum(%x, %y)
}

%add_ {
  %x = f32[] parameter(0)
  %y = f32[] parameter(1)
  ROOT %add = f32[] add(%x, %y)
}

ENTRY %main {
  %arg0.1 = f16[32,131070]{1,0} parameter(0)
  %constant.3 = f16[] constant(-inf)
  %pad1 = f16[32,131070] pad(%arg0.1, %constant.3), padding=0_0_0x0_0_0
  %reduce.8 = f16[32]{0} reduce(%pad1, %constant.3), dimensions={1}, to_apply=%max_
  %broadcast.9 = f16[32,131070]{1,0} broadcast(%reduce.8), dimensions={0}
  %subtract.131070 = f16[32,131070]{1,0} subtract(%arg0.1, %broadcast.9)
  %exponential.11 = f16[32,131070]{1,0} exponential(%subtract.131070)
  %convert.12 = f32[32,131070]{1,0} convert(%exponential.11)
  %constant.13 = f32[] constant(0)
  %pad2 = f32[32,131070] pad(%convert.12, %constant.13), padding=0_0_0x0_0_0
  %reduce.18 = f32[32]{0} reduce(%pad2, %constant.13), dimensions={1}, to_apply=%add_
  %convert.19 = f16[32]{0} convert(%reduce.18)
  %broadcast.20 = f16[32,131070]{1,0} broadcast(%convert.19), dimensions={0}
  ROOT %divide.21 = f16[32,131070]{1,0} divide(%exponential.11, f16[32,131070]{1,0} %broadcast.20)
}
```

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc [hunk 1]
```diff
@@ -1956,26 +1956,26 @@ static void UnrollInnerTileLoop(
     bool check_x_tile_bounds, int64 x_num_steps, int64 step_x,
     int64 vector_size, const string& loop_name, KernelSupportLibrary* ksl,
     llvm::Value* start_offset_x, llvm::Value* y_loc, llvm::Value* tile_width,
-    IrArray::Index& source_idx, llvm::IRBuilder<>& b_,
```
Comment (akuegel) [inline tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc]: Please replace b_ with b

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc [hunk 2]
```diff
@@ -1956,26 +1956,26 @@ static void UnrollInnerTileLoop(
...
+    const IrArray::Index& source_idx, llvm::IRBuilder<>* b_,
     const IrEmitterUnnested::EmitElementFunction* emit_elem_function) {
   llvm::Type* index_ty = tile_width->getType();
   auto constant = [&](int64 val) {
     return llvm::ConstantInt::get(index_ty, val);
   };
-  for (int j = 0; j < x_num_steps / vector_size; j++) {
-    for (int i = 0; i < vector_size; i++) {
+  for (int64 j = 0; j < x_num_steps / vector_size; j++) {
+    IrArray::Index source_idx_x_base =
```
Comment (akuegel) [inline tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:1966]: I think this can be moved outside of the loop. Note that AddOffsetToDim creates a new index and doesn't modify the existing one.

**Review by akuegel — CHANGES_REQUESTED** [review]
Please make sure to address this part of the comment I gave on the old PR:

... because it is failing a DCHECK when running the test tensorflow/compiler/xla/client/lib:arithmetic_test_gpu without -c opt (i.e. with NDEBUG enabled).
Check failed: shape.IsArray() (s32[3], s32[3])

All the review comments I did were things we had already applied internally when trying to land this PR, but this test was still failing a DCHECK in some unrelated file (not changed in this PR) and I didn't immediately see why. I don't see anything in your new code to address this :)

**Comment by nouiz** [comment]
I have many problems to just run the tests you gave. I have errors likes:

/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/Scrt1.o: In function `_start':
(.text+0x20): undefined reference to `main'
collect2: error: ld returned 1 exit status

or like:

2020-04-02 15:47:09.542705: F tensorflow/compiler/xla/tests/client_library_test_base.cc:48] Non-OK-status: result.status() status: Not found: no platforms found could not create local client for testing
external/bazel_tools/tools/test/test-setup.sh: line 310: 44350 Aborted                 (core dumped) "${TEST_PATH}" "$@" 2>&1

or like: 

ERROR: /home/fbastien/github/tensorflow/tensorflow/compiler/xla/client/BUILD:182:1: Linking of rule '//tensorflow/compiler/xla/client:sharding_builder' failed (Exit 1)
gcc: error: missing argument to '-B'


I have this with all tests. So it looks like something is broken upstream (could be specific to my environment). I have errors on the upstream commit before mine too.

Can you gave me the command you use to reproduce the error you have and on which commit you tested it? It is a rebased version of this PR or this PR directly?

I'll continue to try to reproduce your error, but I already spend a few hours.

**Comment by nouiz** [comment]
2 days ago, a build problem was merged upstream that cause the error I saw. I made an issues: https://github.com/tensorflow/tensorflow/issues/38171

Now, hopefully, I can start to look at the problem you mentioned.

**Comment by nouiz** [comment]
After one day I'm still not able to reproduce the error. Are you able to reproduce it? Which bazel command line allow you to reproduce it?

Also, can you give me the full error message? It would be useful to know which DCHECK failed. I do not have that information in this PR or the previous one.

**Review by cheshire — COMMENTED** [review]
LGTM, but can we have a better title and description? I think it's fine to duplicate and link to the previous PR

**Comment by akuegel** [comment]
> After one day I'm still not able to reproduce the error. Are you able to reproduce it? Which bazel command line allow you to reproduce it?

I ran this on f7b584639f07dcc95a405c857ce4071c93b5c9ac (my fixed version of your PR). However I only ran it internally, without specifying -c opt as I normally do so that DCHECK is actually evaluated. I am not so familiar with open source to know how you can make sure that DCHECKs are triggered. The easiest option for you might be to temporarily change the DCHECK in shape_util.cc to a CHECK (see stack trace below to see which line this is).

> Also, can you give me the full error message? It would be useful to know which DCHECK failed. I do not have that information in this PR or the previous one.

Here is the full stack trace:

F0403 00:03:26.839476    1674 shape_util.cc:426] Check failed: shape.IsArray() (s32[3], s32[3])
*** Check failure stack trace: ***
    @     0x7f9965176fcc  absl::logging_internal::LogMessage::DieIfFatal()
    @     0x7f9965176893  absl::logging_internal::LogMessage::SendToLog()
    @     0x7f9965175b3e  absl::logging_internal::LogMessage::Flush()
    @     0x7f9965177e68  absl::logging_internal::LogMessageFatal::~LogMessageFatal()
    @     0x7f9977a50770  xla::ShapeUtil::ElementsIn()
    @     0x7f99889179bf  xla::gpu::(anonymous namespace)::ComputeMaxUnrollFactor()
    @     0x7f998891ffd2  xla::gpu::IrEmitterUnnested::EmitTargetElementLoop()
    @     0x7f99888e2798  xla::gpu::IrEmitter::HandleReduce()
    @     0x7f998891a487  xla::gpu::IrEmitterUnnested::HandleReduce()
    @     0x7f9977f81546  xla::HloInstruction::Visit<>()
    @     0x7f9977f67e50  xla::PostOrderDFS<>()
    @     0x7f9977f80ccc  xla::HloInstruction::Accept<>()
    @     0x7f9989ed69f1  xla::HloComputation::Accept<>()
    @     0x7f99890d7aa1  xla::gpu::GpuCompiler::RunBackend()
    @     0x7f99819ed2ef  xla::LLVMCompiler::Compile()
    @     0x7f998971611e  xla::Service::BuildExecutables()
    @     0x7f9989718f56  xla::Service::ExecuteGraphParallel()
    @     0x7f9989cba4d6  xla::Client::ExecuteParallel()
    @     0x7f9989cb86ff  xla::Client::Execute()
    @     0x7f9989cb80a9  xla::Client::ExecuteAndTransfer()
    @     0x7f9989e7cd74  xla::ClientLibraryTestBase::ExecuteAndTransfer()
    @     0x7f9989e7da3d  xla::ClientLibraryTestBase::ComputeAndCompareLiteralWithStatus()
    @     0x7f9989e7d2a9  xla::ClientLibraryTestBase::ComputeAndCompareLiteral()
    @     0x7f9989ee16fb  xla::ClientLibraryTestBase::ComputeAndCompareR1<>()
    @     0x7f9989ed58fa  xla::(anonymous namespace)::ArithmeticTest_ArgMinR2Axis0_Test::TestBody()
    @     0x7f99661097b4  testing::internal::HandleSehExceptionsInMethodIfSupported<>()
    @     0x7f99660f62f2  testing::internal::HandleExceptionsInMethodIfSupported<>()
    @     0x7f99660e6353  testing::Test::Run()
    @     0x7f99660e69f7  testing::TestInfo::Run()
    @     0x7f99660e704c  testing::TestSuite::Run()
    @     0x7f99660ef335  testing::internal::UnitTestImpl::RunAllTests()
    @     0x7f996610e7b4  testing::internal::HandleSehExceptionsInMethodIfSupported<>()

**Comment by akuegel** [comment]
> After one day I'm still not able to reproduce the error. Are you able to reproduce it? Which bazel command line allow you to reproduce it?

Sorry that you had to spend so much time on this :-(
I didn't expect that this behaves so much differently in open source than internally. I should have provided the full stack trace immediately.
So here is how you can reproduce it in open source:
Change line 426 in tensorflow/compiler/xla/shape_util.cc from DCHECK to CHECK (the line that checks that shape is an array).
Then you should see the CHECK fail.
There might be a way to enable DCHECKs in open source, but I don't know how :-(

**Comment by nouiz** [comment]
I'm not able to have this XLA test works with upstream TF. So I can't try to reproduce this error. 
I created an issues that use Google Docker images: https://github.com/tensorflow/tensorflow/issues/38205

I just pushed a few small changes too. Local tests still compiling...

**Comment by nouiz** [comment]
I added a commit that fix the NCHECK failure.

**Review by akuegel — DISMISSED** [review]
Thanks for figuring out how to fix this. Looks good :)

**Comment by cheshire** [comment]
@nouiz I see small but noticeable regression on this input HLO: https://gist.github.com/cheshire/a7e5e546d5feafdc3db97ef1d89baa91, could you take a look?

**Comment by nouiz** [comment]
Ok. Any idea on which GPU this was run?

**Comment by cheshire** [comment]
Titan V.

**Comment by nouiz** [comment]
I have been able to identify a fusion that give me a speed up of 0.85x:
```
HloModule cluster_0

%add_f32 (x.3360: f32[], y.3361: f32[]) -> f32[] {
  %x.3360 = f32[] parameter(0)
  %y.3361 = f32[] parameter(1)
  ROOT %add.3362 = f32[] add(f32[] %x.3360, f32[] %y.3361)
}

%fused_computation.1884 (param_0.8573: f32[2,384,1024], param_1.10177: f32[2,384]) -> (f32[2,384], f32[2,384,1024], f32[2,384,1024]) {
  %param_0.8573 = f32[2,384,1024] parameter(0)
  %param_1.10177 = f32[2,384] parameter(1)
  %constant_2237_clone_1 = f32[] constant(0.0009765625)
  %broadcast.3809.clone.1 = f32[2,384] broadcast(f32[] %constant_2237_clone_1), dimensions={}
  %multiply.2547.clone.1 = f32[2,384] multiply(f32[2,384] %param_1.10177, f32[2,384] %broadcast.3809.clone.1)
  %broadcast.1940.clone.1 = f32[2,384,1024] broadcast(f32[2,384] %multiply.2547.clone.1), dimensions={0,1}
  %subtract.415 = f32[2,384,1024] subtract(f32[2,384,1024] %param_0.8573, f32[2,384,1024] %broadcast.1940.clone.1)
  %multiply.2546 = f32[2,384,1024] multiply(f32[2,384,1024] %subtract.415, f32[2,384,1024] %subtract.415)
  %constant_2236 = f32[] constant(0)
  %reduce.1694 = f32[2,384] reduce(f32[2,384,1024] %multiply.2546, f32[] %constant_2236), dimensions={2}, to_apply=%add_f32
  ROOT %tuple.744 = (f32[2,384], f32[2,384,1024], f32[2,384,1024]) tuple(f32[2,384] %reduce.1694, f32[2,384,1024] %subtract.415, f32[2,384,1024] %broadcast.1940.clone.1)
}

ENTRY %cluster_0__XlaCompiledKernel_true__XlaHasReferenceVars_false__XlaNumConstantArgs_1083__XlaNumResourceArgs_393_.18693 {
  %get-tuple-element.1047 = f32[2,384,1024] parameter(0)
  %get-tuple-element.1046 =  f32[2,384] parameter(1)
  ROOT %fusion.1884 = (f32[2,384], f32[2,384,1024], f32[2,384,1024]) fusion(f32[2,384,1024] %get-tuple-element.1047, f32[2,384] %get-tuple-element.1046), kind=kInput, calls=%fused_computation.1884
}
```

This fusion have a tuple output. The problem is that LLVM didn't vectorize the store for the non reduction output.

The softmax operation generate one fusion with 2 output: the reduction and the exponential. LLVM vectorize the store in that case...

I'll continue to investigate.

**Comment by nouiz** [comment]
Note, we need to be executed with this flag, as we cannot have fusion with tuples in input HLO. Something in the optimization pipeline doesn't like this: `--xla_flags=--xla_disable_all_hlo_passes`

**Comment by nouiz** [comment]
The MOF fusion with the reduction didn't had the alignment information in the store. So LLVM can't vectorize it. I fixed it for this case. But I saw other places that call directly CreateStore. Converting the one I didn't fix could help other cases.

Should be ready to benchmark again.

**Review by cheshire — DISMISSED** [review]

**Comment by akuegel** [comment]
It seems the vectorization test is now failing, maybe it needs a small adjustment?

**Comment by nouiz** [comment]
I tried the tests here and they pass. I rebased this PR and the test speed pass.

Can you re-run the test? The error message doesnt give me any hints of what the problem is.

**Comment by akuegel** [comment]
You probably didn't click through to the Target Log tab which shows what part didn't match:

https://source.cloud.google.com/results/invocations/c0898e06-66ba-46c6-914a-780826ae3d62/targets/%2F%2Ftensorflow%2Fcompiler%2Fxla%2Fservice%2Fgpu%2Ftests:reduction_vectorization_test/log

**Comment by nouiz** [comment]
Thanks. I pushed a fix. It seems to be a problem in some combination of environment/GPUs. But it should be fixed now.

**Review by cheshire — APPROVED** [review]
